### DISCUSSION TITLE
Is it possible to use firejail with Firefox (tar.bz2) downloaded directly from the Mozilla website ?

### DISCUSSION THREAD

**Discussion by firejailaddssecuirty** [netblue30/firejail#4665] [Q&A] [answered]
I am using all apps that firejail has profile for including Firefox. I have installed Firefox from the Ubuntu repos using apt. **If I download the tar.bz2 directly from the Mozilla website & extract the folder in folder in my home how do I run it inside firejail ?** My guess is firecfg will not able to add it. If I download the Nightly will firejail be able to isolate it ?

This is the link for Nightly >>> [https://www.mozilla.org/en-US/firefox/channel/desktop/#nightly](url)

**Comment by rusty-snake** [comment, marked as answer]
https://github.com/netblue30/firejail/wiki/Sandboxing-Binary-Software#mozilla-firefox-opt-install

If you install it into `$HOME` (e.g. `~/.local/opt/nightly`), you need to `ignore exec ${HOME}`, `ignore apparmor` and `whitelist ${HOME}/.local/opt/nightly`.

FYI: https://codeberg.org/rusty-snake/ffx

**Reply by firejailaddssecuirty** [reply]
See the thing is I have downloaded the tar from their website in the past. I never needed to "install" it. There is a firefox launcher inside the extracted folder. You click it and Firefox launches. So you can put the folder wherever you want. So if I want to run it inside firejail is installing it mandatory ?

**Reply by rusty-snake** [reply]
Unpacking (into a directory better than `~/Downloads` such as `~/.local/opt/nigthly`) is what I call "installing" here.

**Reply by firejailaddssecuirty** [reply]
I have added the following to /etc/firefox.profile
```
ignore exec ${HOME}
ignore apparmor
whitelist ${HOME}/.local/opt/firefox
```
I have download the regular version. Not the nightly. I have copied the firefox folder ~/.local/opt/firefox

This is the output
```
$ firejail /home/home/.local/opt/firefox/firefox
Reading profile /etc/firejail/firefox.profile
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/firefox-common.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Warning: networking feature is disabled in Firejail configuration file
Parent pid 37290, child pid 37291
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Warning: cleaning all supplementary groups
Warning: cleaning all supplementary groups
Warning: cleaning all supplementary groups
Warning: cleaning all supplementary groups
Post-exec seccomp protector enabled
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Child process initialized in 63.81 ms
Exec failed with error: Permission denied

Parent is shutting down, bye...
```

**Reply by rusty-snake** [reply]
> I have added the following to /etc/firefox.profile
> 
> ```
> ignore exec ${HOME}
> ignore apparmor
> whitelist ${HOME}/.local/opt/firefox
> ```

1. Use locals
2. *no*exec
3. If you add it at the end, it has no effect.

**Reply by firejailaddssecuirty** [reply]
Sorry I don't understand what you mean. I have never tried to edit any firejail profile before. What exactly should I do ?

**Reply by firejailaddssecuirty** [reply]
@rusty-snake 
Okay I did it. I was supposed to add those rules under "# Persistent local customizations". One last thing. By adding those 3 rules did I deliberately lose some isolation ?  If yes then what exactly did I compromise ?

**Reply by rusty-snake** [reply]
1. https://github.com/netblue30/firejail/wiki/Creating-overrides
   Create a `firefox.local` file at `~/.config/firejail` (or `/etc/firejail`) containing the three lines from above (and remove them from `/etc/firejail/firefox.profile`).
2. `ignore exec ${HOME}` -> `ignore noexec ${HOME}`
3. `ignore` affects only following command i.e. `apparmor` will be used if `ignore apparmor` is added below of `apparmor`.

**Reply by rusty-snake** [reply]
> By adding those 3 rules did I deliberately lose some isolation ? If yes then what exactly did I compromise ?

- `whitelist ${HOME}/.local/opt/firefox`: obvious but not relevant.
- `ignore noexec ${HOME}` allows executing files in $HOME. (not a isolation command)
- `ignore apparmor` is required too to allow executing files in $HOME. This one can be hardened (drop `ignore apparmor` and instead modify the AA profile) if you use AA but I can not help you there.

**Reply by firejailaddssecuirty** [reply]
I created at file named `firefox.local` inside [~/.config/firejail](url) and added those three lines. Deleted the the lines I added to `/etc/firejail/firefox.profile`. Did a Firefox restart & everything is working. I want to know something. When adding those three line to `/etc/firejail/firefox.profile` worked why create an extra config file name `firefox.local `inside `~/.config/firejail` ? What complications can arise if I directly edit the file `/etc/firejail/firefox.profile` ?

**Reply by rusty-snake** [reply]
> When adding those three line to /etc/firejail/firefox.profile worked why create an extra config file name firefox.local inside ~/.config/firejail ? What complications can arise if I directly edit the file /etc/firejail/firefox.profile ?

https://github.com/netblue30/firejail/wiki/Creating-overrides:

> Modifying a profile file (`.profile`) in `/etc/firejail` is discouraged. Changes will be overwritten when you update Firejail.

**Comment by ghost** [comment]
> 3. ignore apparmor is required too to allow executing files in $HOME. This one can be hardened (drop ignore apparmor and instead modify the AA profile) if you use AA but I can not help you there.

Maybe I can help with the above. Assuming your Firefox 'installation' is under 
${HOME}/.local/opt/firefox, the below snippet needs to be added to `/etc/apparmor.d/local/firejail-default`:
```

# allow running Firefox installed under ${HOME}/.local/opt/firefox
owner @{HOME}/.local/opt/firefox/** ix,
owner @{HOME}/.mozilla/native-messaging-hosts/** ix,
```

Once you've done so you can drop `ignore apparmor` from your `firefox.local` and convince AppArmor to drop its existing caches and account for the newly added rules. The latter needs a `sudo rm -fr <AA-cache-path>/*` plus a reboot, so be careful and double-check your AA's setup, especially if you use a custom cache path. If not, <AA-cache-path> defaults to using /var/cache/apparmor or /etc/apparmor.d/cache (depending on versioning).
